staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -359,18 +359,30 @@ type WaitFunc func(done <-chan struct{}) <-chan struct{}
 // ErrWaitTimeout will be returned if the channel is closed without fn ever
 // returning true.
 func WaitFor(wait WaitFunc, fn ConditionFunc, done <-chan struct{}) error {
-	c := wait(done)
+	stopCh := make(chan struct{})
+	once := sync.Once{}
+	closeCh := func() {
+		once.Do(func() {
+			close(stopCh)
+		})
+	}
+	defer closeCh()
+	c := wait(stopCh)
 	for {
-		_, open := <-c
-		ok, err := fn()
-		if err != nil {
-			return err
-		}
-		if ok {
-			return nil
-		}
-		if !open {
-			break
+		select {
+		case _, open := <-c:
+			ok, err := fn()
+			if err != nil {
+				return err
+			}
+			if ok {
+				return nil
+			}
+			if !open {
+				return ErrWaitTimeout
+			}
+		case <-done:
+			closeCh()
```

Comment:
I believe this is wrong. I do not think this will break you out of the for loop, which I believe is the desired behavior. If we just return here, that will have us exit the for loop and then the deferred close will be called. Otherwise we are relying on waiting for the timeout which seems wrong. We should make sure we are properly testing this case.

Reply:
If we decide that `WaitFor` should stop waiting once `done` channel is closed, regardless of the `wait` func, then I suggest that we apply @cheftako's comment in a different PR, because that's a behavior change while this PR is fixing a goroutine leak.

Reply:
I'd be ok with creating an issue to make sure this does not get dropped. Also we should properly document the parameters when we make that change.

Reply:
As @caesarxuchao said, It just goes to another loop and running the final test. 

This behavior is described in the documents.
```go
// WaitFor gets a channel from 'wait()'', and then invokes 'fn' once for every value
// placed on the channel and once more when the channel is closed.
```

Reply:
Nvm, created https://github.com/kubernetes/kubernetes/issues/72357.